### PR TITLE
Update to new NumPy RNG

### LINKED ISSUE #4273: Move to more recent numpy random generator
### What is the expected enhancement?
We make use of a lot of `np.random.RandomState` generators, however these are now legacy code.  It has been replaced with `np.random.default_rng` that generates different values.  We should move to this new generator. Ideally before the circuit library comes out because the seed values, and the offset, can be (should be) used to uniquely identify the circuits.  This makes that identification a bit more future proof.

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary

fixes #4273 

We used `RandomState` a lot, but that is now legacy code.  This moves us to `np.random.default_rng`:

https://numpy.org/doc/stable/reference/random/generator.html?highlight=default_rng#numpy.random.default_rng

that makes use of `PCG64`:

https://numpy.org/doc/stable/reference/random/bit_generators/pcg64.html#numpy.random.PCG64

Besides removing legacy code, this is also 4x faster at initializing the generators.

### CODE REVIEW

**Comment by manoelmarques** [comment]
I believe numpy min.requirements need to change to >=1.17 right ? I need to change Aqua RandomState usage too.

**Comment by nonhermitian** [comment]
This is a good point.

File: qiskit/quantum_info/states/random.py [hunk 1]
```diff
@@ -117,18 +117,18 @@ def _ginibre_matrix(nrow, ncol, seed):
     Args:
         nrow (int): number of rows in output matrix.
         ncol (int): number of columns in output matrix.
-        seed(int or RandomState): RandomState for rng.
+        seed(int or default_rng): default_rng for rng.
```
Comment (ajavadia) [inline qiskit/quantum_info/states/random.py:120]:
```suggestion
        seed(int or np.random.Generator): default_rng for rng.
```

can you do a grep for `"or default_rng"` and change all of them

Comment (nonhermitian) [reply qiskit/quantum_info/states/random.py]: This is done.

File: qiskit/quantum_info/operators/random.py [hunk 2]
```diff
@@ -64,18 +64,18 @@ def random_hermitian(dims, traceless=False, seed=None):
         traceless (bool): Optional. If True subtract diagonal entries to
                           return a traceless hermitian operator
                           (Default: False).
-        seed (int or RandomState): Optional. Set a fixed seed or
+        seed (int or default_rng): Optional. Set a fixed seed or
```
Comment (chriseclectic) [inline qiskit/quantum_info/operators/random.py:67]: Should this be
```suggestion
        seed (generator_seed or Generator): ...
```
(and same for all the other locations)
Looking at the documentation for the new `Generator` it accepts a lot of different things for seed values:
```
seed : {None, int, array_like[ints], ISeedSequence, BitGenerator, Generator}, optional
        A seed to initialize the `BitGenerator`. 
```

Comment (nonhermitian) [reply qiskit/quantum_info/operators/random.py:67]: But we do not need to use all of them.  That seems outside of this PR.

File: qiskit/quantum_info/states/random.py (hunk 1 above)
Comment (ajavadia) [reply qiskit/quantum_info/states/random.py]: @nonhermitian I still see "or default_rng" appearing in some places.

File: qiskit/quantum_info/operators/random.py [hunk 3]
```diff
@@ -36,18 +36,18 @@ def random_unitary(dims, seed=None):
...
                                    generator for RNG.
 
     Returns:
         Operator: a unitary operator.
     """
     if seed is None:
-        random_state = np.random
-    elif isinstance(seed, RandomState):
+        random_state = np.random.default_rng()
+    elif isinstance(seed, np.random.Generator):
```
Comment (chriseclectic) [inline qiskit/quantum_info/operators/random.py:47]: The reason for the `None` case before was to bypass the slow initialization of `RandomState`. This shouldn't be needed now with `default_rng` so all these if blocks can be simplified to:

```python
if isinstance(seed, np.random.Generator):
    random_state = seed
else:
    random_state = np.random.default_rng(seed)
```

Comment (chriseclectic) [reply qiskit/quantum_info/operators/random.py:47]: If we wanted to speed up these functions even more we could also leave the `None` case and assign it to a global variable `default_rng` initialied on import of the random.py module.

ie.
```python
DEFAULT_RNG = default_rng()

def  random_unitary(dims, seed=None):
    if seed is None:
        random_state = DEFAULT_RNG
    elif...
```

Comment (nonhermitian) [reply qiskit/quantum_info/operators/random.py:47]: Your more than welcome to in another PR.  This one was fun enough.

**Comment by chriseclectic** [comment]
@nonhermitian You will need to update the minimum Numpy and Scipy requirements to `scipy>=1.4, numpy>=1.17` (I was running scipy 1.3.1 locally and the `scipy.states.unitary_group` function didn't work with the new generators).

**Comment by nonhermitian** [comment]
Ok missed the scipy requirement

File: qiskit/quantum_info/operators/random.py (hunk 3 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/random.py:47]: Out of interest I timed these two cases (initializing new default each function call, vs using a module level default rng) and got this:

rng_init

File: qiskit/quantum_info/operators/random.py (hunk 2 above)
Comment (jaygambetta) [reply qiskit/quantum_info/operators/random.py]: i agree this is future pr

File: qiskit/quantum_info/operators/random.py (hunk 3 above)
Comment (jaygambetta) [reply qiskit/quantum_info/operators/random.py:47]: this is nice but its a new pr

**Pull request merged**